### DISCUSSION TITLE
Support exporting figures to separate files

### DISCUSSION THREAD

**Discussion by multimeric** [quarto-dev/quarto-cli#6702] [Feature Requests]
### Description

Some (all?) journals expect you to submit each figure individually, so it would be lovely for Quarto to do this for us, given that it knows about each figure. I just had to manually copy and paste 20 figures from a word document as part of a submission process which was a bit painful.

[This has already been discussed in the context of knitr code chunks](https://github.com/quarto-dev/quarto-cli/discussions/4254), but I'm after a more general mechanism, that will work with any `{#fig-XXX}` block including images, un-executed code etc.

I would envisage this working as a new output format, such as `pdf-figures`, such that `quarto render --to pdf-figures` would dump all figures as `figure_1.pdf`, `figure_2.pdf` etc.

**Comment by cscheid** [comment]
This kind of use-case is precisely the reason for the existence of Lua filters in Pandoc (and quarto). We don't intend (nor have the resources) to implement every use case like this, but you should be able to easily extract those figures, especially with the new crossref work that should be landing this week.

See https://github.com/quarto-dev/quarto-web/blob/07bc9caa09a9b1627b94012b988af071c52e03f2/docs/prerelease/1.4/lua_changes.qmd, specifically the `FloatRefTarget` node, for what will be possible.

**Reply by multimeric** [reply]
Thanks! I had to some extent come to this conclusion myself. The API updates in 1.4 will make this even more pleasant.

However, in trying to implement this as a filter, I was struck with two issues:
* Neither pandoc nor quarto support making multiple outputs from one input document. This makes it impossible to solve this in the ideal way.
* Once we take a figure out of the context of the document, it loses its figure number because it lacks the preceding figures. At least I think this is true.

**Reply by cscheid** [reply]
> Neither pandoc nor quarto support making multiple outputs from one input document

I'm not sure what you mean by this, but Lua is a general-purpose programming language. You can write to files, and you can call any other programs from it you want.

> Once we take a figure out of the context of the document, it loses its figure number because it lacks the preceding figures. At least I think this is true.

Don't take them out of the context of the document, but instead create a copy of the relevant files in a location of your choice.

**Reply by multimeric** [reply]
> Don't take them out of the context of the document, but instead create a copy of the relevant files in a location of your choice.

I'm not sure what you mean by this. Could you elaborate?

In any case, I've written a very simple Quarto plugin that solves this use case for me. However, I have significant issues with LaTeX based formats, as I describe in the readme: https://github.com/multimeric/QuartoSplitFigures

**Reply by cscheid** [reply]
> > Don't take them out of the context of the document, but instead create a copy of the relevant files in a location of your choice.
>
> I'm not sure what you mean by this. Could you elaborate?

This 

```lua

local image_output_path
return {
  {
    Meta = function(meta)
      image_output_path = meta.image_output_path or "image_outputs"
      pandoc.system.make_directory(image_output_path)
    end
  },
  {
    Image = function(image)
      local fin = io.open(image.src, "rb")
      local fout = io.open(pandoc.path.join(image_output_path, image.src), "w")
      fout:write(fin:read("*all"))
      fout:close()
    end
  }
}
```

I might be missing something (and I haven't tested that filter at all), but this makes a copy of your images. Another simpler option that occurred to me right now is to just use `keep-tex: true`, which will keep the intermediate .tex file and all other files around.

**Reply by multimeric** [reply]
Yes I understand that, and that's very similar to my extension's solution. However I want to capture the entire figure div, which may involve code instead of images, and also including the caption. This isn't as trivial, because the figure numbering in the LaTeX engine is based on the number of figures in the document, so copying the figure div as I do currently will break the numbering.

**Reply by cscheid** [reply]
Oh, you need the actual numbering? The LaTeX numbering is resolved by the LaTeX engine at the time of LaTeX compilation, so what you're attempting to do is fundamentally impossible in Lua. You'll need to find a tool that operates directly on the resulting PDF. There's unfortunately nothing that Quarto or Pandoc can do to help here.
